docs/e2e_testing.md

```diff
@@ -0,0 +1,398 @@
+# Automated end-to-end testing
+
+## Introduction
+
+From this tutorial, you will get to know how to create an e2e test for a workflow.
+Automated e2e testing can help you develop workflows faster and avoid errors.
+
+As an example, you will create an e2e test for the workflow that calculates aggregates on the `bigquery-public-data:crypto_bitcoin.transactions` BigQuery table. 
+The workflow calculates:
+
+* Daily transaction count
+* Daily transaction fee sum 
+
+The workflow saves the result in a BigQuery table.
+
+You will write tests for two implementations of that workflow. The first implementation 
+utilizes BigQuery (as the input and the output) and Dataflow (for processing).
+
+The second implementation utilizes only BigQuery, for both io and processing.
+
+Workflows using BigQuery are interesting to test because there is no way emulate 
+BigQuery on a local machine.
+
+## Preparation
+
+For this tutorial, you need a fresh BigFlow project.
+
+First, create a directory where you want to store the example project:
+
+```shell script
+mkdir ~/bigflow_cookbook
+cd ~/bigflow_cookbook
+```
+
+Now, prepare a virtual environment for the project. To do so, follow the BigFlow installation guide. The virtual environment should be placed inside 
+the `~/bigflow_cookbook` directory.
+
+Next, [create a new BigFlow project](scaffold.md) called "btc_aggregates":
+
+```shell script
+bf start-project
+```
+
+Install the project requirements:
+
+```shell script
+cd ~/bigflow_cookbook/btc_aggregates_project
+pip install -r resources/requirements.txt
+```
+
+Finally, remove examples from the generated project:
+
+```shell script
+cd ~/bigflow_cookbook/btc_aggregates_project
+rm -rf btc_aggregates/internationalports
+rm -rf btc_aggregates/wordcount
+```
+
+## Testing Dataflow + BigQuery implementation
+
+Take a look at the workflow implementation below. The important part (in the context of e2e testing) of that workflow is the configuration. Also, save the following code as a module inside the generated project: `btc_aggregates/btc_aggregates_df_bq.py`.
```

Comment:
This code should be generated by bigflow start-project and should not require creation bt user

Reply:
I don't want to put it into the scaffold to avoid additional maintenance.

Reply:
Putting additional code in docs also puts some additional maintenance ;) And because it is out of the main code it is harder to maintain imo. So maybe we should unify existing scaffold with e2e example?

Reply:
I have added the following note:

This tutorial uses `bigflow==1.1.3`

Now it requires pretty much zero maintenance. And it's still fine, even if there is a new bigflow version, like `1.6.8`, that tutorial still provides value because it shows what's the general idea behind e2e testing in bigflow. And at the same time, it's not a critical part of bigflow.

The only reasons I can see for adding this example to the scaffold are:

- We will avoid some incompatibilities with newer versions (though that tutorial should be ok with any version below 2.0).
- It's a bit easier to just generate the project and have the example inside the generated project.

The scaffold is a way to start a project, and not a place to store code for the tutorials.

@anjensan give us your thoughts, please

Reply:
The continuation of that discussion will be in Andreis comment.